proposals/webhooks-push.md

```diff
@@ -0,0 +1,177 @@
+
+This proposal adds the ability to push events from the EntityEvent Resource using webhooks.
+
+# Rationale
+Rather than polling on the EntityEvent Resource for updates, it's useful for data producers to push events to consumer APIs. 
+
+# Proposal
+When an _EntityEventSequenceNumeric_, _ResourceName_, _ResourceRecordKey_ tuple has been published to the EntityEvent resource, providers MAY "push" events to data consumers for replication or other event sourcing applications. 
+
+In doing so, the replication process becomes simpler for the consumer, as they only need to receive EntityEvents from the producer and pick up the corresponding records from provider APIs. In general, this is preferable to having to poll continuously for new records.
+
+# Examples
+The following examples show using the EntityEvent model along with [RESO Common Format](reso-common-format.md) , [OAuth2 bearer tokens](https://oauth.net/2/bearer-tokens/), and HTTP POST requests.
+
+RESO Common Format supports both single and multi-valued payloads.
+
+## 1. Producer Sends a Single EntityEvent to the Consumer API
+
+**REQUEST**
+```
+POST https://example.api.com/webhooks
+  Content-Type: application/json
+  Authorization: Bearer <token>
+```
+```json
+{
+  "@reso.context": "urn:reso:metadata:2.0:resource:entityevent",
+  "EntityEventSequence": 1101,
+  "ResourceName": "Showing",
+  "ResourceRecordKey": "235021"
+}
+```
+
+**RESPONSE**
+```
+HTTP/2 200
+Content-Type: application/json
+```
+
+## 2. Producer Sends a Batch of EntityEvent Records to the Consumer API
+
+**REQUEST**
+```
+POST https://example.api.com/webhooks
+  Content-Type: application/json
+  Authorization: Bearer <token>
+```
+```json
+{
+  "@reso.context": "urn:reso:metadata:2.0:resource:entityevent",
+  "value": [{
+    "EntityEventSequence": 1101,
+    "ResourceName": "Property",
+    "ResourceRecordKey": "235021"
+  }, {
+    "EntityEventSequence": 1111,
+    "ResourceName": "Media",
+    "ResourceRecordKey": "ABC123"
+  }]
+}
+```
+
+**RESPONSE**
+```
+HTTP/2 200
+Content-Type: application/json
+```
+
+## 3. Producer Sends a Batch of EntityEvent Records to a Consumer API With Retry-After
+
+**REQUEST**
+```
+POST https://example.api.com/webhooks
+  Content-Type: application/json
+  Authorization: Bearer <token>
+```
+```json
+{
+  "@reso.context": "urn:reso:metadata:2.0:resource:entityevent",
+  "value": [{
+    "EntityEventSequence": 1101,
+    "ResourceName": "Property",
+    "ResourceRecordKey": "235021"
+  }, {
+    "EntityEventSequence": 1111,
+    "ResourceName": "Media",
+    "ResourceRecordKey": "ABC123"
+  }]
+}
+```
+
+**RESPONSE**
+```
+HTTP/2 429
+Content-Type: application/json
+Retry-After: 3600
+```
+
+## 4. Producer Sends a Batch of EntityEvent Records to a Consumer API with Optional EntityEventSource Header
+
+**REQUEST**
+```
+POST https://example.api.com/webhooks
+  Content-Type: application/json
+  Authorization: Bearer <token>
+  EntityEventSource: IDX Feed 123
```

Comment:
@psftc - this used to be called "ConsumerLabel" but it's too generic of a term in my opinion. Is it OK to change? This was a minor part of the spec, but people could potentially using it ... they could still do so though.

Reply:
Yes, I think it is sensible to make this change.